### ISSUE TITLE
ShortCodeParser: A parameter value of 0 is ignored by the parser and fails.

### ISSUE THREAD

**Issue by Josua2012** [silverstripe/silverstripe-framework#2382]
This type of shortcode:
[SendaContent id='0']
produces the following error:
```
[Notice] Undefined offset: 2
GET /

Line 222 in C:\wamp\www\framework\parsers\ShortcodeParser.php
Source

213 
214 				// Pull the attributes out into a key/value hash
215 				$attrs = array();
216 				
217 				if (!empty($match['attrs'][0])) {
218 					preg_match_all(self::attrrx(), $match['attrs'][0], $attrmatches, PREG_SET_ORDER);
219 
220 
221 					foreach ($attrmatches as $attr) {
222 						list($whole, $name, $value) = array_values(array_filter($attr));
223 						$attrs[$name] = $value;
224 				    }
225 
226 				}
227 				
228 				// And store the indexes, tag details, etc

Trace

    ShortcodeParser->extractTags([SendaContent id='0'])
    ShortcodeParser.php:367
    ShortcodeParser->replaceElementTagsWithMarkers([SendaContent id='0'])
    ShortcodeParser.php:522
    ShortcodeParser->parse([SendaContent id='0'])
    SendaConfig.php:1455
    SendaConfig::ContentParser([SendaContent id='0'],Page)
    SendaPageDecorator.php:247
    SendaPageDecorator->UpdateContentAndSearch(,,,,,,)
    Object.php:989
    Object->extend(UpdateContentAndSearch,)
    Page.php:12
    Page->UpdateContentAndSearch()
    SendaConfig.php:1892
    {closure}()
    call_user_func(Closure)
    Fluent.php:465
    Fluent::with_locale(en_US,Closure)
    SendaConfig.php:1897
    SendaConfig::CheckOperations()
    SendaConfig.php:1512
    SendaConfig::Init(Page_Controller)
    Page.php:39
    Page_Controller->init()
    Controller.php:139
    Controller->handleRequest(SS_HTTPRequest,DataModel)
    ContentController.php:220
    ContentController->handleRequest(SS_HTTPRequest,DataModel)
    ModelAsController.php:68
    ModelAsController->handleRequest(SS_HTTPRequest,DataModel)
    FluentRootURLController.php:95
    FluentRootURLController->handleRequest(SS_HTTPRequest,DataModel)
    Director.php:325
    Director::handleRequest(SS_HTTPRequest,Session,DataModel)
    Director.php:143
    Director::direct(/,DataModel)
    main.php:189
```
This occurs for this code: 
function extractTags($content) (ShortcodeParser.php)
```
list($whole, $name, $value) = array_values(array_filter($attr));
```
since the variable $attr contains the following array:
```
$attr [3]
  0 = "id='0'"
  1 = "id"
  2 = "0"
```
And the following code modifies the array
```
array_values(array_filter($attr));
```
Now the array contains only two elements:
```
0 = "id='0'"
1 = "id"
```
which makes the error occurs when assigned to
```
list($whole, $name, $value)
```

Please, fix this problem as soon as possible.

Thanks,
Regards,
Jose A.

**Comment by td204** [comment]
This issue is still around in 3.1.15: the fix in https://github.com/silverstripe/silverstripe-framework/pull/2417 works for me.

**Comment by sheadawson** [comment]
This is an issue for me still also

**Comment by hafriedlander** [comment]
#2417 still needs a fix before it can be merged - as it is, it fixes a subset of the larger issue (`="0"`) but not the whole issue (`=""` would still break). If someone wants to find some time to fix the PR, the principal behind it seems fine.

**Comment by tractorcow** [comment]
I'll see about fixing this up for the 2.2.2 release (and 3.3.0)

**Comment by dhensby** [comment]
Bump

**Comment by tractorcow** [comment]
I'm not gonna have time I think. :) Too late to get this into 3.3.0.

**Comment by dunatron** [comment]
This issue is still around

![screen shot 2016-03-04 at 12 15 47 pm](https://cloud.githubusercontent.com/assets/8181248/13512994/f3be4bae-e202-11e5-832a-1e09dce9f186.png)

**Comment by Firesphere** [comment]
I'll have a look for SS4, seems to be the too broad empty check.

**Comment by tractorcow** [comment]
It was fixed with https://github.com/silverstripe/silverstripe-framework/commit/8ae794ee99b17bed83d969005be03a6621a4ccad#diff-419fde14ad0518fd3b25838354c25674L265 in 4.x.

**Comment by dhensby** [comment]
Fixed in 3.2+ with https://github.com/silverstripe/silverstripe-framework/pull/5761

Note that @hafriedlander's example edgecase of `[shorcode id=""]` actually returns `array('id' => '""')` so I didn't address it as that's maybe kind of indented maybe probably?
